ch_7_Finetune/7_1_Finetune_Llama2-7B.md

```diff
@@ -0,0 +1,321 @@
+A popular open-source LLM [meta-llama/Llama-2-7b-hf](https://huggingface.co/meta-llama/Llama-2-7b-hf) is chosen to illustrate the process of model finetuning.
+
+> **Note**
+>
+> You can specify the argument `pretrained_model_name_or_path` with both Huggingface repo id or local model path.
+> If you have already downloaded the Llama 2 (7B) model, you could specify `pretrained_model_name_or_path` to the local model path.
+
+With BigDL-LLM optimization, you can load the model with `bigdl.llm.transformers.AutoModelForCausalLM` instead of `transformers.AutoModelForCausalLM` to conduct implicit quantization.
+
+For Intel GPUs, once you have the model in low precision, **set it to `to('xpu')`**.
+
+```python
+model = AutoModelForCausalLM.from_pretrained(pretrained_model_name_or_path = "meta-llama/Llama-2-7b-hf",
+                                            load_in_low_bit="nf4",
+                                            optimize_model=False,
+                                            torch_dtype=torch.float16,
+                                            modules_to_not_convert=["lm_head"])
+model = model.to('xpu')
+
+```
+
+> **Note**
+>
+> We specify load_in_low_bit="nf4" here to apply 4-bit NormalFloat optimization. According to the [QLoRA paper](https://arxiv.org/pdf/2305.14314.pdf), using "nf4" could yield better model quality than "int4".
+
+### 7.1.2.2 Prepare Model for Training
+Then we apply `prepare_model_for_kbit_training` from `bigdl.llm.transformers.qlora` to preprocess the model for training. 
+
+```python
+from bigdl.llm.transformers.qlora import prepare_model_for_kbit_training
+model.gradient_checkpointing_enable()
+model = prepare_model_for_kbit_training(model)
+```
+
+Next, we can obtain a PEFT model from the optimized model and a configuration object containing the parameters as follows:  
+
+```python
+from bigdl.llm.transformers.qlora import get_peft_model
+from peft import LoraConfig
+
+config = LoraConfig(r=8, # defines how many parameters will be trained
+                    lora_alpha=32, # adjusts the magnitude of the weight matrix
+                    target_modules=["q_proj", "k_proj", "v_proj"], 
+                    lora_dropout=0.05, # used to avoid overfitting
+                    bias="none", 
+                    task_type="CAUSAL_LM")
+model = get_peft_model(model, config)
+
+```
+> **Note**
+>
+> Instead of `from peft import prepare_model_for_kbit_training, get_peft_model` as we did for regular QLoRA using bitandbytes and cuda, we import them from `bigdl.llm.transformers.qlora` here to get a BigDL-LLM compatible PEFT model. And the rest is just the same as regular LoRA finetuning process using `peft`.
+>
+
+> **Note**
+>
+> For Llama 2 (7B) model, the `target_modules` is set as `["q_proj", "k_proj", "v_proj"]`.
+> You can find a list of **target_modules available** on the [Hugging Face Documentation](https://github.com/huggingface/peft/blob/39ef2546d5d9b8f5f8a7016ec10657887a867041/src/peft/utils/other.py#L220).
```

Comment:
Is there any reason to specify to this specific commit?

Reply:
Just follow another comprehensive tutorial to write this part ( Should I remove it?